### ISSUE TITLE
Use BigQuery Dataframes as Connector to BigQuery

### ISSUE THREAD

**Issue by OELSJAN** [pola-rs/polars#17326]
### Description

BigQuery launched a feature named "[BigQuery Dataframes](https://cloud.google.com/python/docs/reference/bigframes/latest)":

> BigQuery DataFrames provides a Pythonic DataFrame and machine learning (ML) API powered by the BigQuery engine.

With this you can execute Pandas operations directly on BigQuery engine. So maybe this API can be used to implement a better connector towards BigQuery, which also supports some lazy optimizations like filter pushdown, instead of using `from_arrow` with a hardcoded query executed by the Python BigQuery client.

**Comment by tswast** [comment]
Thanks @OELSJAN for the feature request! I'm curious to hear if there's anything we can do in BigQuery DataFrames to make these filter pushdown features easier to implement.

Also, for your awareness, we have on our backlog a request to make an official polars connector for BigQuery (watch issue https://github.com/googleapis/python-bigquery/issues/1979 for updates). I suspect that might be a good place to implement such optimizations, as a separate package could make dependencies a little easier to manage.

See also this request on the BigQuery DataFrames repo for some polars support. https://github.com/googleapis/python-bigquery-dataframes/issues/735, which is mostly focused on the I/O piece similar to this request.

**Comment by tswast** [comment]
A question: are there other I/O methods that support push-down to the storage layer? I'm curious what hooks are available for such functionality.

**Comment by alexander-beedie** [comment]
> A question: are there other I/O methods that support push-down to the storage layer? I'm curious what hooks are available for such functionality.

@tswast: Yes - for example, the Polars Iceberg integration supports various pushdown optimisations including predicates, range queries, and suchlike. Might be helpful to look at that[^1] 👍

I've been meaning to look at integrating the BigQuery Client object as a valid connection type for our `pl.read_database` function for a while... should really get around to it!

[^1]: feat(python): Add support for Iceberg (https://github.com/pola-rs/polars/pull/10375)

**Comment by tswast** [comment]
> Yes - for example, the Polars Iceberg integration

Very cool!

From what I've heard from folks (e.g. https://github.com/googleapis/python-bigquery/issues/1979), it's important to avoid unnecessary dependencies on pyarrow, so it's good to see that [`_scan_python_function`](https://github.com/pola-rs/polars/blob/106e23936bd8496210c08df76f6bf7b679bac86c/py-polars/polars/lazyframe/frame.py#L336) has an option to avoid that dependency.

How stable is this interface? I'm curious if this sort of connector is best contributed directly to the polars package or should be provided by a separate package (similar to how pandas has refactored BigQuery support out into pandas-gbq years ago).

Since this proposed feature is to go beyond predicates to potentially turning aggregates and such into BigQuery queries, maybe it's best to stick to the pola-rs/polar repo for now, as that scan functionality improves/extends?

**Comment by tswast** [comment]
I've started a [BigQuery + polars gist](https://gist.github.com/tswast/99b017b20386e324f5c7d2bd49f21b5f) with some ideas. I'll try to keep that up to date as I experiment with reads and writes. The first experiment, [bigquery-to-polars-no-pyarrow-ipynb](https://gist.github.com/tswast/99b017b20386e324f5c7d2bd49f21b5f#file-bigquery-to-polars-no-pyarrow-ipynb), is a barebones read API that doesn't require `pyarrow` to go from BigQuery table -> polars DataFrame.

This could be extended further to support a `scan_bigquery` with no extra dependencies (such as bigframes), as the BigQuery Storage Read API supports predicates (see `row_restriction` in https://cloud.google.com/python/docs/reference/bigquerystorage/latest/google.cloud.bigquery_storage_v1.types.ReadSession.TableReadOptions). Maybe I'll give that a try soon, as it sounds like a fun project. :-)

**Comment by OELSJAN** [comment]
I am really overwhelmed by the stone i have set rolling here and i am delighted that this topic is now being adressed. Can't wait to process bigquery data in a more native way.

**Comment by tswast** [comment]
I'm making some good progress on a `scan_bigquery()` function that avoids a pyarrow dependency in https://github.com/tswast/polars/blob/issue17326-scan-bigquery/py-polars/polars/io/bigquery.py

Still quite a few TODOs left (actually fetch the schema and convert to polars schema instead of hardcoding the test table, actually apply the predicates, actually apply the column filters). I'm feeling confident with this approach, though.

Note: I'm targeting a `scan_bigquery()` function first since it could actually be re-used for the `pl.read_database` (https://github.com/pola-rs/polars/issues/17326#issuecomment-2328866738) code. Every query job (generated by `Client.query(...)`) in BigQuery actually generates a destination table that could be read from in the same way. Aside: `Client.query_and_wait()` might not create a table for small results, so there is some subtlety there, but TBD if it's worth it to include such optimizations)

**Comment by leonelqz** [comment]
Hi! 

Excellent initiative @tswast !  I would be a frequent user of this feature. Having data stored in BigQuery and doing processing using polars is a great idea!
